lib/lookdevXUsd/UsdMxVersionUpgrade.cpp

```diff
@@ -0,0 +1,1097 @@
+using _DownstreamInputPortList = std::vector<std::pair<UsdShadeConnectableAPI, UsdShadeInput>>;
+
+_DownstreamOutputPortList _getDownstreamOutputPorts(UsdShadeConnectableAPI node)
+{
+    // Returns a list(node, port) of all items that connect to the output port of the node passed in as parameter.
+    //    
+    // We assume a nicely behaved graph without connections teleporting across NodeGraph boundaries.
+    _DownstreamOutputPortList retVal;
+    auto ng = UsdShadeConnectableAPI(node.GetPrim().GetParent());
+    if (!ng) {
+        return retVal;
+    }
+    // Look for NodeGraph connections:
+    for (const auto& output : ng.GetOutputs())
+    {    
+        for (const auto& sourceInfo : ng.GetConnectedSources(output))
+        {
+            if (sourceInfo.source.GetPrim() == node.GetPrim())
+            {
+                retVal.push_back(std::make_pair(ng, output));
+            }
+        }
+    }
+    return retVal;
+}
+
+_DownstreamInputPortList _getDownstreamInputPorts(UsdShadeConnectableAPI node)
+{
+    // Returns a list(node, port) of all items that connect to the output port of the node passed in as parameter.
+    //    
+    // We assume a nicely behaved graph without connections teleporting across NodeGraph boundaries.
+    _DownstreamInputPortList retVal;
+    auto ng = UsdShadeConnectableAPI(node.GetPrim().GetParent());
+    if (!ng) {
+        return retVal;
+    }
+    // Check every node inside the graph:
+    for (const auto& child : ng.GetPrim().GetChildren())
+    {
+        auto shader = UsdShadeShader(child);
+        if (!shader) {
+            continue;
+        }
+        for (const auto& input : shader.GetInputs()) {
+            for (const auto& sourceInfo : shader.ConnectableAPI().GetConnectedSources(input)) {
+                if (sourceInfo.source.GetPrim() == node.GetPrim()) {
+                    retVal.push_back(std::make_pair(shader.ConnectableAPI(), input));
+                }
+            }
+
+        }
+    }
+
+    return retVal;
+}
+
+void _moveInput(UsdShadeShader sourceNode, const TfToken& sourceInputName, UsdShadeShader destNode, const TfToken& destInputName)
+{
+    auto editor = UsdNamespaceEditor(sourceNode.GetPrim().GetStage());
+    editor.ReparentProperty(sourceNode.GetPrim().GetAttribute(UsdShadeUtils::GetFullName(sourceInputName, UsdShadeAttributeType::Input)),
+                            destNode.GetPrim(), UsdShadeUtils::GetFullName(destInputName, UsdShadeAttributeType::Input));
+    if (editor.CanApplyEdits()) {
+        editor.ApplyEdits();
+    } else {
+        TF_WARN("Failed to move input '%s' from node '%s' to input '%s' on node '%s'. Please make sure the material layer is writable.",
+                sourceInputName.GetText(),
+                sourceNode.GetPrim().GetPath().GetText(),
+                destInputName.GetText(),
+                destNode.GetPrim().GetPath().GetText());
```

Comment:
If the layer is locked, or the material is in a Payload, the namespace editor will not be able to modify the data. Probably a good idea.
If the material is in a simple unlocked sublayer that is not the current edit layer, then the namespace editor will still be able to modify the original data. We will want to document that as destructive operations appear in the material layer while additive operations appear as new opinions on the edit layer, for a maximum amount of confusion.